### PR TITLE
Leaderboard

### PR DESCRIPTION
**Pull request by eluric**
Added a command to show the leaderboard of the current teams. Ties between teams will be broken by the latest correct submission time. E.g Team A and B have both solved the same amount of puzzles but team B will show up higher than A because the latest puzzle they solved was at an earlier time than the latest puzzle A solved.

Also fixed the issue for submission where the number of puzzles a team solved was not updated in the public.teams table.

### CODE REVIEW

File: cogs/puzzle.py [hunk 1]
```diff
@@ -66,6 +106,8 @@ async def submit_answer(
 
         if not submission_is_correct:
             return await interaction.followup.send("The submitted answer is incorrect!")
+        else:
+            await increase_puzzles_solved(player.team_name)
```
Comment (SkellyBG) [inline cogs/puzzle.py:110]: don't need an else here - the if statements returns c:

File: cogs/puzzle.py [hunk 2]
```diff
@@ -22,10 +23,49 @@
...
+        )
+
+
 class Puzzle(commands.GroupCog):
     def __init__(self, bot):
         self.bot = bot
 
+    def update_leaderboard(self, new_embeds: list[discord.Embed]):
+        self.leaderboard_embeds = new_embeds
+
```
Comment (SkellyBG) [inline cogs/puzzle.py:68]: I can't find where this function is used

File: src/queries/team.py [hunk 3]
```diff
@@ -101,3 +101,38 @@ async def remove_team(team_name: str):
...
+    puzzles_solved = (await acur.fetchone())[0]
+
+    await acur.execute(
+        """
+        UPDATE public.teams
+        SET puzzle_solved = %s
+        WHERE team_name = %s
+        """,
+        (puzzles_solved + 1, team_name),
+    )
```
Comment (SkellyBG) [inline src/queries/team.py:134]: You can do something like this :D

```suggestion
    await acur.execute(
        """
        UPDATE public.teams
        SET puzzle_solved = puzzle_solved + 1
        WHERE team_name = %s
        """,
        (puzzles_solved + 1, team_name),
    )
```

File: cogs/puzzle.py [hunk 4]
```diff
@@ -22,10 +23,49 @@
 HUNT_START_TIME: datetime = config["HUNT_START_TIME"]
 
 
+class PaginationView(View):
```
Comment (SkellyBG) [inline cogs/puzzle.py:26]: :partying_face: AMAZING WORK

**Review by SkellyBG — APPROVED** [review]
few minor nits, but amazing work as usual Maria!

File: src/queries/puzzle.py [hunk 5]
```diff
@@ -128,3 +128,26 @@ async def delete_puzzle(puzzle_id: str):
...
+
+    await acur.execute(
+        """
+        SELECT t.team_name, t.puzzle_solved
+        FROM public.teams AS t LEFT JOIN public.submissions AS s
+        ON (t.team_name = s.team_name)
+        AND s.submission_is_correct = TRUE
+        GROUP BY t.team_name
+        ORDER BY t.puzzle_solved DESC, MAX(s.submission_time) ASC
+        """
```
Comment (SkellyBG) [inline src/queries/puzzle.py:145]: in the case that the puzzle_solved and submission_time is the same, we should sort on team_name as well (only ever likely for when teams have 0 solves)

File: cogs/puzzle.py (hunk 2 above)
Comment (eluric) [reply cogs/puzzle.py:68]: forgot to remove that

File: src/queries/team.py (hunk 3 above)
Comment (eluric) [reply src/queries/team.py:134]: except i wouldn't have the puzzles_solved + 1 in the tuple, right

File: src/queries/puzzle.py (hunk 5 above)
Comment (eluric) [reply src/queries/puzzle.py:145]: oh yeah maybe. I didn't think it mattered for the order of teams with no solves but I could do that.

File: src/queries/team.py (hunk 3 above)
Comment (SkellyBG) [reply src/queries/team.py:134]: oh wait yup, good catch

**Pull request merged**